tests/test_create_layout.py

```diff
@@ -0,0 +1,206 @@
+      'one.tgz': '1234567890aabbcc',
+      'foo/two.tgz': '0000001111112233',
+      'three.txt': '1111222233334455',
+      'bar/bat/four.tgz': '6677889900123456'
+    }
+
+    unchanged_artifacts, modified_artifacts, added_artifacts, deleted_artifacts = \
+        create_layout.changes_between_snapshots(self.before, after)
+
+    self.assertSetEqual(unchanged_artifacts, self.empty_set)
+    self.assertSetEqual(modified_artifacts,
+        {'bar/bat/four.tgz', 'foo/two.tgz', 'one.tgz', 'three.txt'})
+    self.assertSetEqual(added_artifacts, self.empty_set)
+    self.assertSetEqual(deleted_artifacts, self.empty_set)
+
+
+  def test_partially_modified_filesystem_snapshot(self):
+
+    after = {
+      'five.txt': '5555555555555555',
+      'one.tgz': '1234567890abcdef',
+      'foo/two.tgz': 'ffffffffffffffff',
+      'bar/bat/four.tgz': '6677889900123456',
+      'baz/six.tgz': '6666666666666666'
+    }
+
+    unchanged_artifacts, modified_artifacts, added_artifacts, deleted_artifacts \
+        = create_layout.changes_between_snapshots(self.before, after)
+
+    self.assertSetEqual(unchanged_artifacts, {'one.tgz'})
+    self.assertSetEqual(modified_artifacts, {'bar/bat/four.tgz', 'foo/two.tgz'})
+    self.assertSetEqual(added_artifacts, {'baz/six.tgz', 'five.txt'})
+    self.assertSetEqual(deleted_artifacts, {'three.txt'})
+
+
+  def test_create_material_rules_of_initial_step(self):
+    # Zero index means that the current step is the initial step,
+    # so we need to ALLOW all the existing files instead of matching.
+    second_link = in_toto.models.link.Link.read(self.second_step_link_str)
+    links = [second_link]
+
+    expected_materials = [
+      ['ALLOW', 'bar/bat/four.tgz'],
+      ['ALLOW', 'one.tgz'],
+      ['ALLOW', 'foo/two.tgz'],
+      ['ALLOW', 'three.txt'],
+      ['DELETE', 'three.txt'],
+      ['DISALLOW', '*']
+    ]
+
+    # create_material_rules returns a lists with random order,
+    # so we need to sort the lists
+    self.assertEqual(sorted(expected_materials),
+        sorted(create_layout.create_material_rules(None, second_link)))
+
+  def test_create_material_rules_of_not_initial_step(self):
+    # Nonzero index means that the current step is not the initial step,
+    # so we need to MATCH materials with products of the previous step.
+    first_link = in_toto.models.link.Link.read(self.first_step_link_str)
+    second_link = in_toto.models.link.Link.read(self.second_step_link_str)
+    links = [first_link, second_link]
+
+    expected_materials = [
+      ['MATCH', 'bar/bat/four.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'one.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'foo/two.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'three.txt', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['DELETE', 'three.txt'],
+      ['DISALLOW', '*']
+    ]
+
+    # create_material_rules returns a lists with random order,
+    # so we need to sort the lists
```

Comment:
That is a problem. The order of a list of artifact rules does matter. For instance, if you call `'DISALLOW' *` first, then every subsequent rule is moot. 

I think we should order the artifacts before we iterate over them in `create_material_rules` and `create_product_rules`.

Reply:
Sorry, I didn't say clear enough here. The rules still follow the correct order, but the order of artifacts inside each rule is random. For example, in a material rule, the order of matching different artifacts is random. But this material rule still keeps all the "MATCH" rules the first, and the "DISALLOW *" rule the last.

Or do you mean the order of artifacts inside each rule is also important? What I concern about is that sorting the artifacts will take extra time.

Reply:
Understood. And no, I don't think the order of artifacts inside each group is important. This is more a problem of the tests, which don't verify if `MATCH` rules are first, `DISALLOW *` is last, etc., right? And it is also a problem of determinism. It would be nice to get the same results across different Python versions. 

I wouldn't worry too much about the extra time ordering will take (let's cross the performance optimization bridge when we get there).

Reply:
I have implemented the sort. Please check it. Thanks! @lukpueh